smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/TypeScriptJmesPathVisitor.java

```diff
@@ -0,0 +1,315 @@
+import software.amazon.smithy.jmespath.ast.IndexExpression;
+import software.amazon.smithy.jmespath.ast.LiteralExpression;
+import software.amazon.smithy.jmespath.ast.MultiSelectHashExpression;
+import software.amazon.smithy.jmespath.ast.MultiSelectListExpression;
+import software.amazon.smithy.jmespath.ast.NotExpression;
+import software.amazon.smithy.jmespath.ast.ObjectProjectionExpression;
+import software.amazon.smithy.jmespath.ast.OrExpression;
+import software.amazon.smithy.jmespath.ast.ProjectionExpression;
+import software.amazon.smithy.jmespath.ast.SliceExpression;
+import software.amazon.smithy.jmespath.ast.Subexpression;
+
+class TypeScriptJmesPathVisitor implements ExpressionVisitor<Void> {
+
+    TypeScriptWriter writer;
+    String executionContext;
+
+    String accessor;
+    JmespathExpression jmesExpression;
+    int scopeCount;
+
+    TypeScriptJmesPathVisitor(TypeScriptWriter writer, String accessor, JmespathExpression expression) {
+        this.writer = writer;
+        this.accessor = accessor;
+        this.executionContext = accessor;
+        this.jmesExpression = expression;
+        scopeCount = 0;
+    }
+
+    private String makeNewScope(String prefix) {
+        scopeCount += 1;
+        return prefix + scopeCount;
+    }
+
+    private static boolean isNumeric(String strNum) {
+        if (strNum == null) {
+            return false;
+        }
+        try {
+            Double.parseDouble(strNum);
+        } catch (NumberFormatException nfe) {
+            return false;
+        }
+        return true;
+    }
+
+    public void run() {
+        writer.openBlock("let returnComparer = () => {", "}", () -> {
+            executionContext = this.accessor;
+            jmesExpression.accept(this);
+            writer.write("return $L;", this.executionContext);
+        });
+        executionContext = "returnComparer()";
+    }
+
+
+    void writeBooleanExpectation(String expectedValue, String returnValue) {
+        writer.openBlock("if ($L == $L) {", "}", this.executionContext, expectedValue, () -> {
+            writer.write("return $L", returnValue);
+        });
+    }
+
+    void writeAnyStringEqualsExpectation(String expectedValue, String returnValue) {
+        String element = makeNewScope("element");
+        writer.openBlock("for (let $L of $L) {", "}", element, this.executionContext, () -> {
+            writer.openBlock("if ($L == $L) {", "}", element, expectedValue, () -> {
+                writer.write("return $L;", returnValue);
+            });
+        });
+    }
+
+    void writeAllStringEqualsExpectation(String expectedValue, String returnValue) {
+        String element = makeNewScope("element");
+        String result = makeNewScope("result");
+        writer.write("let $L = false;", result);
+        writer.openBlock("for (let $L of $L) {", "}", element, this.executionContext, () -> {
+            writer.write("$L = $L || ($L == $L)", result, result, element, expectedValue);
```

Comment:
Hm, all string equals means every string must equal value. So shouldn't this be `&&` not `||`?

```
$L = $L || ($L == $L)
```

Seems to be an any string equals. Plugging in some values here. Say the expectedValue is "a", and the values being checked are "a" and "b", then you have:

```
result = result || "a" == "a"; // true
result = result || "b" == "a"; // true, but should be false.
```

Reply:
nice catch!

Reply:
added 
```
writer.write("let $L = ($L.length > 0);", result, this.executionContext);
```
to start as false the array is empty to start

Reply:
Resolving as it adheres to https://github.com/awslabs/smithy/pull/664